src/main/java/software/amazon/event/kafkaconnector/batch/BestEffortInOrderEventBridgeBatching.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package software.amazon.event.kafkaconnector.batch;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collector;
+import java.util.stream.Stream;
+import software.amazon.awssdk.services.eventbridge.model.PutEventsRequestEntry;
+import software.amazon.event.kafkaconnector.util.MappedSinkRecord;
+
+public class BestEffortInOrderEventBridgeBatching implements EventBridgeBatchingStrategy {
+
+  private final Collector<
+          MappedSinkRecord<PutEventsRequestEntry>,
+          PutEventsRequestEntrySizeCollector,
+          Stream<List<MappedSinkRecord<PutEventsRequestEntry>>>>
+      collector =
+          Collector.of(
+              PutEventsRequestEntrySizeCollector::new,
+              PutEventsRequestEntrySizeCollector::accumulate,
+              (left, right) -> {
+                throw new IllegalStateException();
+              },
+              PutEventsRequestEntrySizeCollector::finish);
+
+  @Override
+  public Stream<List<MappedSinkRecord<PutEventsRequestEntry>>> apply(
+      Stream<MappedSinkRecord<PutEventsRequestEntry>> records) {
+    if (records.isParallel()) {
+      throw new IllegalArgumentException("Stream must not be parallel.");
+    }
+    return records.collect(collector);
+  }
+
+  static class PutEventsRequestEntrySizeCollector {
+
+    private static final int MAX_BATCH_SIZE = 256 * 1024;
+
+    private final List<List<MappedSinkRecord<PutEventsRequestEntry>>> result = new ArrayList<>();
+    private int actualBatchSize = 0;
+    private int index = 0;
+
+    PutEventsRequestEntrySizeCollector() {
+      result.add(new ArrayList<>());
+    }
+
+    void accumulate(final MappedSinkRecord<PutEventsRequestEntry> item) {
+
+      int itemSize = getSize(item.getValue());
+      if (actualBatchSize + itemSize > MAX_BATCH_SIZE) {
+        result.add(new ArrayList<>());
+        index++;
+        actualBatchSize = 0;
+      }
+
+      actualBatchSize += itemSize;
+      result.get(index).add(item);
+    }
+
+    Stream<List<MappedSinkRecord<PutEventsRequestEntry>>> finish() {
+      return ((result.size() == 1) && (actualBatchSize == 0)) ? Stream.empty() : result.stream();
```

Comment:
Can you please explain what this does for me, i.e. when is `finish` called in a stream?

Reply:
The collector will call `finish` after all elements of the given stream are collected/processed. The first empty batch is added on initialization (constructor) see comment above. If there is no element added it would return an empty batch. Maybe I should document it.